src/main/java/com/sucy/skill/listener/ShieldBlockListener.java

```diff
@@ -0,0 +1,43 @@
+package com.sucy.skill.listener;
+
+import com.sucy.skill.api.event.PlayerBlockDamageEvent;
+import org.bukkit.Bukkit;
+import org.bukkit.Statistic;
+import org.bukkit.entity.Player;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.entity.EntityDamageByEntityEvent;
+import org.bukkit.event.player.PlayerStatisticIncrementEvent;
+
+import java.util.HashMap;
+import java.util.Timer;
+import java.util.TimerTask;
+
+/**
+ * Listener to throw custom {@link PlayerBlockDamageEvent}
+ */
+public class ShieldBlockListener extends SkillAPIListener {
+
+    private final HashMap<Player, EntityDamageByEntityEvent> volatileMap = new HashMap<>();
+
+    /**
+     * Remember EntityDamageByEntityEvent in a moment
+     */
+    @EventHandler
+    public void onDamage(EntityDamageByEntityEvent event){
+        if (!(event.getEntity() instanceof Player)) return;
+        Player player = (Player) event.getEntity();
+        if (!player.isBlocking()) return;
+        volatileMap.put(player, event);
+    }
+
+    /**
+     * Trigger for {@link PlayerBlockDamageEvent}
+     */
+    @EventHandler
+    public void onStat(PlayerStatisticIncrementEvent event){
+        if (!event.getStatistic().equals(Statistic.DAMAGE_BLOCKED_BY_SHIELD)) return;
+        Player player = event.getPlayer();
+        if (!volatileMap.containsKey(player)) return;
+        Bukkit.getPluginManager().callEvent(new PlayerBlockDamageEvent(volatileMap.get(player), event));
```

Comment:
We should probably remove the player from the volatile map here so we don't cause the map to potentially grow indefinitely. There may be some potential for some race conditions here, but I think the risk is probably fairly low?

Reply:
I would argue that storing large amounts of events won't drain as much performance as constantly adding and removing them. I'm thinking of periodically deleting events. However, an exception exists if the event is still in the waiting phase

Reply:
I would argue that storing large amounts of events won't drain as much performance as constantly adding and removing them. I'm thinking of periodically deleting events. However, an exception exists if the event is still in the waiting phase

Reply:
> There may be some potential for some race conditions here
It's really disturbing

Reply:
I'm less concerned about performance in the short-term and more concerned with the effects having a server on for several days with a lot of unique players. First off, it has been stated by Spigot/Bukkit developers before that storing player objects in collections is not wise as the players themselves hold a fair amount of data and by holding a reference to them, it doesn't allow the garbage collector to clean them up, which can cause plenty of issues. The other aspect is storing the event itself which holds plenty of reference to various other things.

We should use UUIDs as the key instead of the player, and we should make sure we're cleaning up upon completion. It may cause a very slight performance hit to add/remove all the time, but it wouldn't cause a compounding issue in the long run.

Reply:
Okay, Reduce storage and bring timer back